Production/plugins/TrainTupleProducer.cc

```diff
@@ -0,0 +1,694 @@
+#include "DataFormats/JetReco/interface/CaloJetCollection.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/RecoCandidate/interface/RecoEcalCandidate.h"
+#include "DataFormats/TauReco/interface/PFTau.h"
+#include "DataFormats/TauReco/interface/PFTauDiscriminator.h"
+#include "DataFormats/TauReco/interface/PFTauTransverseImpactParameter.h"
+#include "DataFormats/TauReco/interface/PFTauTransverseImpactParameterAssociation.h"
+#include "DataFormats/TauReco/interface/PFTauTransverseImpactParameterFwd.h"
+#include "DataFormats/TrackReco/interface/HitPattern.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+
+#include "FWCore/Common/interface/TriggerNames.h"
+#include "FWCore/Framework/interface/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+
+#include "AnalysisDataFormats/TopObjects/interface/TtGenEvent.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+#include "CommonTools/Utils/interface/StringToEnumValue.h"
+#include "RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h"
+#include "SimDataFormats/GeneratorProducts/interface/GenEventInfoProduct.h"
+
+#include "TauMLTools/Analysis/interface/GenLepton.h"
+#include "TauMLTools/Analysis/interface/SummaryTuple.h"
+#include "TauMLTools/Analysis/interface/TauIdResults.h"
+#include "TauMLTools/Analysis/interface/TrainTuple.h"
+#include "TauMLTools/Core/interface/Tools.h"
+#include "TauMLTools/Core/interface/TextIO.h"
+#include "TauMLTools/Production/interface/GenTruthTools.h"
+#include "TauMLTools/Production/interface/MuonHitMatch.h"
+#include "TauMLTools/Production/interface/TauAnalysis.h"
+#include "TauMLTools/Production/interface/TauJet.h"
+
+#include "Geometry/CaloGeometry/interface/CaloCellGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+#include "Geometry/CaloTopology/interface/CaloTowerTopology.h"
+#include "Geometry/CaloTopology/interface/CaloTowerConstituentsMap.h"
+#include "Geometry/CaloTopology/interface/HcalTopology.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+#include "RecoLocalCalo/CaloTowersCreator/interface/EScales.h"
+#include "RecoLocalCalo/CaloTowersCreator/src/CaloTowersCreator.h"
+#include "RecoLocalCalo/EcalRecAlgos/interface/EcalSeverityLevelAlgoRcd.h"
+
+namespace tau_analysis {
+
+struct TrainTupleProducerData {
+    using clock = std::chrono::system_clock;
+
+    const clock::time_point start;
+    train_tuple::TrainTuple trainTuple;
+    train_tuple::AuxPathTuple auxTuple;
+    tau_tuple::SummaryTuple summaryTuple;
```

Comment:
Why do we need summary and aux tuple? I propose to store all global quantities in a `L2SummaryTuple`.